File: winter2009/comp2150/a3/q1/SoopaStore.java
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;

/**
 * Simulates a supermarket. 
 *
 * @input Should be instantiated with a number of checkouts to
 * simulate and an input file to read from
 *
 * @output Prints a summary of most events as they happen and some
 * basic stats at the end.
 */
public class SoopaStore {

  private static Checkout    checkouts[]; ///< Array of checkouts
  private static int         numCheckouts; ///< Number of checkouts this store has
  private static OrderedList events; ///< Sorted linked list of upcoming events
  private static Scanner     inputScanner; ///< Used to parse file input
  private static OrderedList cashierList; ///< List of departed cashiers
  private static int         mostRecentTime;
  
  /**
   * The Main loop essentially. Read in an event, then as long as the
   * event list has more events to process, process them.
   */
  public static void startSimulation() {
    readNextArrival();
    while (! events.isEmpty()) {
      processNextEvent();
    }
  }

  public static int currentTime() {
    if (! events.isEmpty()) {
      Event evt = (Event) events.peekFirst();
      return evt.getTime();
    } else {
      return mostRecentTime;
    }
  }

  /** Pull an event off the EventList and call action() on it to run it. */
  public static void processNextEvent() {
    Event evt = (Event) events.removeFirst();
    mostRecentTime = evt.getTime();
    evt.action();
  }
  
  /** Return this store's event list. */
  public static OrderedList getEventList() {
    return events;
  }

  /**
   * Return the first checkout that doesn't have a cashier, so
   * employees can figure out where to go
   */
  public static Checkout firstUnmannedCheckout() {
    Checkout first = null;
    for (int i=0; i<numCheckouts; ++i) {
      if (!checkouts[i].hasCashier()) {
        first = checkouts[i];
        break;
      }
    }
    return first;
  }
  
  /** Return the emptiest checkout, so customers can figure out where to go */
  public static Checkout preferableCheckout() {
    Checkout pref=null;
    int min = -1; // A value that will never be returned naturally.
                  // We'll overwrite this regardless of the result of getLength().
    int length;
    for (int i=0; i<numCheckouts; ++i) {
      if (checkouts[i].isOpen()) {
        length = checkouts[i].getLength();
        if (min == -1 || length < min) {
          min = length;
          pref = checkouts[i];
        }
      }
    }
    return pref;    
  }
  
  /** true if the input isn't all used up. Are there more people coming? */
  private static boolean hasMoreArrivals() {
    // Logically, hasNextLine makes more sense, but if there's a blank
    // line at the end of the file, hasNextLine will cause errors.
    // We're guaranteed lines start with an integer, so this works better.
    return (inputScanner.hasNextInt());
  }


  /**
   * Pull the next arrival from input, create an event for it, and
   * stick it on the EventList
   */
  public static void readNextArrival() {
    String personType;
    int arrivalTime, employeeNumber, quittingTime, numItems;

    if (hasMoreArrivals()) {
      arrivalTime = inputScanner.nextInt();
      personType  = inputScanner.next();
      if (personType.equals("E")) {
        // This is an employee.
        employeeNumber = inputScanner.nextInt();
        quittingTime   = inputScanner.nextInt();
        addCashierArrivalEvent(arrivalTime,employeeNumber,quittingTime);
      } else {
        // This is a customer.
        numItems = inputScanner.nextInt();
        addCustomerArrivalEvent(arrivalTime,numItems);
      }
    }

  }

  /**
   * Create an ArrivalEvent for an employee record that was pulled from input
   * @param arrivalTime the time at which the person arrives
   * @param employeeNumber id of employee
   * @param quittingTime time at which employee's shift is over
   */
  public static void addCashierArrivalEvent(int arrivalTime, int employeeNumber, int quittingTime) {
    Cashier employee = new Cashier(employeeNumber, quittingTime, arrivalTime);
    CashierArrivalEvent event = new CashierArrivalEvent(arrivalTime, employee);
    events.insertInOrder(event);
  }

  /**
   * Create an ArrivalEvent for a customer record that was pulled from input
   * @param arrivalTime the time at which the person arrives
   * @param numItems the number of items this customer is purchasing
   */
  public static void addCustomerArrivalEvent(int arrivalTime, int numItems) {
    Customer customer = new Customer(numItems, arrivalTime);
    CustomerArrivalEvent event = new CustomerArrivalEvent(arrivalTime, customer);
    events.insertInOrder(event);
  }

  /**
   * Start the simulation by calling SoopaStore.
   * @param args not like we use it.
   */
  public static void main(String args[]) {
    String infile;
    Scanner sc;

    numCheckouts = 8;
    events       = new OrderedList();
    cashierList  = new OrderedList();
    
    // Get input filename.
    sc = new Scanner(System.in);
    System.out.println("Please enter input file name:");
    infile = sc.next();
    sc.close();

    checkouts = new Checkout[numCheckouts];
    for (int i=0; i<numCheckouts; ++i) {
      checkouts[i] = new Checkout(i);
    }
    try {
      inputScanner = new Scanner(new File(infile));
    } catch (FileNotFoundException fnfe) {
      System.out.println(fnfe);
      System.exit(1);
    }
    
    System.out.println("Welcome to SoopaStore!\n");
    
    startSimulation();

    System.out.println("\n...All events complete.  Final Summary:");

    System.out.println("\nCashier Information:\n");
    System.out.println(
      "      ID    Start     Quit        Time    Time    Speed   Customers    Items\n"+
      "  Number     Time     Time      Worked    Idle               Served   Packed\n"+
      "----------------------------------------------------------------------------");
    System.out.printf("%8s %8s %8s %11s %8s %7s %11s %8s\n",
                      0,0,0,0,0,0,0,0);
    

    System.out.println("\nCheckout Information:\n");
    System.out.println(
      "Checkout     Time    Total       Total    Total\n"+
      "  Number  Staffed  Workers   Customers    Items\n"+
      "-----------------------------------------------" );
    for (int i=0; i<numCheckouts; ++i) {
      System.out.printf("%8s %8s %8s %11s %8s\n",
                        i,
                        SoopaStore.checkouts[i].getTimeStaffed(),
                        SoopaStore.checkouts[i].getNumCashiers(),
                        SoopaStore.checkouts[i].getNumCustomers(),
                        SoopaStore.checkouts[i].getTotalItems());
    }
    
    System.out.println("\nCustomers processed in total: ");
    System.out.println("Average waiting time per customer : ");
    
    
    System.out.println("End of Processing");
  }
  
}

File: winter2009/comp2150/a5/q2/lib/soopa_store.rb
# file lib/soopa_store.rb

# Run a simulation of a supermarket.
class SoopaStore

  def initialize
    raise TypeError "don't instantiate this class."
  end

  # ActiveSupport has a method `cattr_reader`. I wish I could use it here.
  def self.checkout;     @@checkout;     end
  def self.events;       @@events;       end
  def self.customers;    @@customers;    end
  def self.cashiers;     @@cashiers;     end
  def self.current_time; @@current_time; end

  # Start the simulation Read events from the specified file.
  # Read one at a time, processing events until an Arrival is processed.
  def self.run(infile)

    @@checkout  = Checkout.new
    @@events    = LinkedList.new :sorted => true

    @@customers = LinkedList.new :sorted => false
    @@cashiers  = LinkedList.new :sorted => false
      
    File.open(infile,'r') do |file|
      until file.eof?
        line = file.readline.split /\s+/
        event = case line[1]
        when "C" then CustomerArrivalEvent.new( line[0].to_i, line[2].to_i )
        when "E" then CashierArrivalEvent.new(  line[0].to_i, line[2].to_i )
        end
        @@events.add event unless event.nil?
        process_events_until ArrivalEvent
      end
      process_remaining_events
    end

    print_stats

    puts "\nEnd of Processing"
  end

  private
  # Pull events from the event list until we find one that's an instance of Class klass.
  def self.process_events_until(klass)
    begin
      event = @@events.pop
      @@current_time = event.time
      event.run
    end until event.is_a?(klass)
  end

  # Process all remaining events in the event list.
  def self.process_remaining_events
    until @@events.empty?
      event = @@events.pop
      @@current_time = event.time
      event.run
    end
  end

  # Print a nice formatted output of the results of the simulation.
  def self.print_stats
    puts "...All events complete. Final Summary:\n"
    cashiers = Table.new do |table|
      table.headers("ID Number", "Start Time", "Time Worked", "Customers Served", "Items Packed")
      @@cashiers.each do |c|
        table.data(c.eid,c.arrival_time,c.time_in_store,c.customers_served,c.total_items)
      end
    end
    
    customers = Table.new do |table|
      table.headers("ID Number", "Start Time", "Grocery Items", "Waiting Time")
      @@customers.each do |c|
        table.data(c.customer_id,c.arrival_time,c.num_groceries,c.time_spent_waiting)
      end
    end

    checkouts = Table.new do |table|
      table.headers("Time Staffed", "Total Customers", "Total Items")
      table.data(@@cashiers.map{|c|c.time_in_store}.inject{|a,v|a+v},
                 @@checkout.customers_served,@@checkout.total_items)
    end    

    average_wait_time = @@customers.map{|c|c.time_spent_waiting}.inject{|a,v|a+v} /
      (@@customers.count.to_f)
    
    puts "\nCashier Information:\n\n",
    cashiers,
    "\nCustomer Information:\n\n",
    customers,
    "\nCheckout Information:\n\n",
    checkouts,
    "\nCustomers processed in total: #{@@customers.count}",
    "Average waiting time per customer :#{average_wait_time}"
    
  end
  
end
